tokio/src/sync/notify.rs

```diff
@@ -917,10 +917,17 @@ impl Notified<'_> {
                         }
                     }
 
+                    let mut old_waker = None;
                     if waker.is_some() {
                         // Safety: called while locked.
+                        //
+                        // This code makes the following assignment, but it
+                        // stores the old waker in `old_waker` so that we can
+                        // drop it after releasing the mutex.
+                        //
+                        // (*waiter.get()).waker = waker;
                         unsafe {
-                            (*waiter.get()).waker = waker;
+                            old_waker = std::mem::replace(&mut (*waiter.get()).waker, waker);
```

Comment:
`old_waker` is always `None` here, as this is the first place we store a waker and from here we always transition to `Waiting` with no return.

Reply:
I am reluctant to rely on assumptions like "waker is always None here" or "waker is always Some here", and prefer to program defensively. We've run into trouble with those in the past due to this type of assumption being broken by future changes to the code.

Anyway, I added a comment to note this.